plasmapy/analysis/time_series/conditional_averaging.py

```diff
@@ -0,0 +1,358 @@
+            signal, peak_indices
+        )
+
+        self._conditional_variance = list(
+            self._calculate_conditional_variance(conditional_events)
+        )
+
+        self._conditional_average = list(self._conditional_average)
+
+        if self._astropy_unit is not None:
+            self._peaks *= self._astropy_unit
+            self._conditional_average *= self._astropy_unit
+
+    @property
+    def time(self):
+        """
+        Time values corresponding to the analysis window.
+
+        Returns
+        -------
+        time : 1D |array_like|
+            Time values representing the analysis window.
+        """
+        return self._return_time
+
+    @property
+    def average(self):
+        """
+        Conditional average over events.
+
+        Returns
+        -------
+        average : 1D |array_like|
+            Array representing the conditional average over events.
+
+        """
+        return self._conditional_average
+
+    @property
+    def variance(self):
+        """
+        Conditional variance over events.
+
+        Returns
+        -------
+        variance : 1D |array_like|
+            Array representing the conditional variance over events.
+
+        """
+        return self._conditional_variance
+
+    @property
+    def peaks(self):
+        """
+        Peak values of conditional events.
+
+        Returns
+        -------
+        peaks : 1D |array_like|
+            Peak values of conditional events.
+
+        """
+        return self._peaks
+
+    @property
+    def waiting_times(self):
+        """
+        Waiting times between consecutive peaks.
+
+        Returns
+        -------
+        waiting_times : 1D |array_like|
+            Waiting times between consecutive peaks of conditional events.
+
+        """
+        return self._waiting_times
+
+    @property
+    def arrival_times(self):
+        """
+        Arrival times corresponding to the conditional events.
+
+        Returns
+        -------
+        arrival_times : 1D |array_like|
+            Arrival times the conditional events.
+
+        """
+        return self._arrival_times
+
+    @property
+    def number_of_events(self):
+        """
+        Total number of conditional events.
+
+        Returns
+        -------
+        number_of_events : int
+            Total number of conditional events.
+
+        """
+        return self._number_of_events
+
+    # This astropy unit checks are quite ugly in my view.
+    # If a code reviewer has a better idea how to handle this I would be very grateful.
+    def _check_units_consistency(self, signal, lower_threshold, upper_threshold):
+        if isinstance(signal, u.Quantity):
+            if not isinstance(lower_threshold, u.Quantity):
+                raise TypeError(
+                    "signal/reference_signal and lower_threshold must have same astropy unit"
+                )
+            if signal.unit != lower_threshold.unit:
```

Comment:
One thing we can do with `Quantity` objects is compare their physical types.  That would be something like `signal.unit.physical_type != lower_threshold.unit.physical_type`, which would be `True` if the units were km and cm, for example.  In this case, the code later on requires them to be in the same units.  

Another possibility would be to use a `try`/`except` block:
```Python
try:
    lower_threshold.to(signal.unit)
except u.UnitConversionError:
    raise u.UnitConversionError("<error message>")
```

Reply:
I adopted the `signal.unit.physical_type != lower_threshold.unit.physical_type` approach. This is better than my first implementation since it would have thrown an error for `u.cm` and `u.m`.